### PR TITLE
Fix aspect based proto builds

### PR DESCRIPTION
**Pull request by ianthehat**
This changes the structure of our providers.
It is now much more principled about configuration specific and transitive
providers, separating them cleanly.
It also allows the GoLibray provider to supply a function that can be used to
produce the configuration specific version (GoSource). This allows us to fix the
proto library dependancies when built in aspect mode.

### CODE REVIEW

File: go/private/rules/cgo.bzl [hunk 1]
```diff
@@ -224,29 +227,36 @@ def _pure(ctx, mode):
...
+  source["cgo_archive"] = library.cgo_archive
+  source["runfiles"] = source["runfiles"].merge(attr.codegen.data_runfiles)
+
+
 def _cgo_collect_info_impl(ctx):
   codegen = ctx.attr.codegen[_CgoCodegen]
   runfiles = ctx.runfiles(collect_data = True)
   runfiles = runfiles.merge(ctx.attr.codegen.data_runfiles)
+
+  library = new_go_library(ctx, _cgo_library_to_source,
```
Comment (jayconrod) [inline go/private/rules/cgo.bzl]: Nit: make `resolver` a keyword arg for readability.

File: go/private/rules/test.bzl [hunk 2]
```diff
@@ -84,23 +90,22 @@ def _go_test_impl(ctx):
   )
 
   # Now compile the test binary itself
-  _, goarchive, executable = go_toolchain.actions.binary(ctx, go_toolchain,
+  test_library = new_go_library(ctx, _testmain_library_to_source, srcs=[main_go], importable=False)
```
Comment (jayconrod) [inline go/private/rules/test.bzl]: Nit: make `resolver` a keyword argument for readability.

File: go/private/rules/test.bzl [hunk 3]
```diff
@@ -84,23 +90,22 @@ def _go_test_impl(ctx):
...
+      source = test_source,
       gc_linkopts = gc_linkopts(ctx),
       x_defs=ctx.attr.x_defs,
   )
 
   # TODO(bazel-team): the Go tests should do a chdir to the directory
   # holding the data files, so open-source go tests continue to work
   # without code changes.
-  runfiles = goarchive.runfiles.merge(ctx.runfiles(files = [executable]))
+  runfiles = ctx.runfiles(files = [executable])
```
Comment (jayconrod) [inline go/private/rules/test.bzl]: Pile-on nit: remove TODO above.

File: proto/def.bzl [hunk 4]
```diff
@@ -37,7 +53,12 @@ _go_proto_aspect = aspect(
     attr_aspects = ["deps", "embed"],
 )
 
+def _proto_library_to_source(ctx, attr, source):
+  compiler = attr.compiler[GoProtoCompiler]
+  merge_embed(source, attr.compiler)
```
Comment (jayconrod) [inline proto/def.bzl]: This doesn't seem right. `compiler` is unused at least.

File: go/private/rules/helpers.bzl [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    "sets",
+)
+load("@io_bazel_rules_go//go/private:providers.bzl",
+    "GoLibrary",
+    "GoSource",
+    "GoArchive",
+    "GoAspectProviders",
+)
+
+def new_go_library(ctx, resolver=None, importable=True, **kwargs):
```
Comment (jayconrod) [inline go/private/rules/helpers.bzl]: Need documentation. In particular, what is the type of `resolver`, and what is it supposed to do?

File: go/private/common.bzl [hunk 6]
```diff
@@ -107,10 +107,11 @@ def go_importpath(ctx):
   Returns:
     Go importpath of the library
   """
-  path = ctx.attr.importpath
+  path = getattr(ctx.attr, "importpath", None)
```
[the remark this answers is not in the document]
Comment (ianthehat) [reply go/private/common.bzl]: That was deliberate, things that don't even have an import path attribute don't get an import path calculated, the default value of the attribute is ""

File: go/private/providers.bzl [hunk 7]
```diff
@@ -15,78 +15,35 @@
 load("@io_bazel_rules_go//go/private:mode.bzl", "mode_string")
 
 GoLibrary = provider()
-"""See go/providers.rst#GoLibrary for full documentation."""
-
-GoPackage = provider()
-
-GoPath = provider()
+"""
```
[the remark this answers is not in the document]
Comment (ianthehat) [reply go/private/providers.bzl]: I am wary of adding documentation in multiple places, if skydoc actually worked or providers actually verified their fields I would use them, but without that it's just a random collection of strings that may be out of date with reality, and also with the rst.

File: go/private/rules/helpers.bzl [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  if GoAspectProviders in dep:
+    return dep[GoAspectProviders].source
+  return dep[GoSource]
+
+def get_archive(dep):
+  if GoAspectProviders in dep:
+    return dep[GoAspectProviders].archive
+  return dep[GoArchive]
+
+def merge_embed(source, embed):
```
[the remark this answers is not in the document]
Comment (ianthehat) [reply go/private/rules/helpers.bzl]: Nope, used in anything that implements a resolver, which includes the proto rules.

File: go/private/rules/helpers.bzl (hunk 5 above)
Comment (ianthehat) [reply go/private/rules/helpers.bzl]: Will document in a follow up cl, the interface changes the the toolchain and stdlib cleanup that comes next.

**Pull request merged**